### DISCUSSION TITLE
Having trouble doing native builds on macOS 11 x86_64 runner.

### DISCUSSION THREAD

**Discussion by cheofusi** [pypa/cibuildwheel#1200] [General]
Hi everyone,

I was hoping this [simple workflow](https://github.com/cheofusi/just_playback/blob/main/.github/workflows/wheels.yml) gets me Windows , Linux & MacOS wheels. It works for all except the macOS runner, [failing](https://github.com/cheofusi/just_playback/runs/7547598326?check_suite_focus=true) with a  `OSError: [Errno 66] Directory not empty`.  Seems to me like an issue with setuptools on the macOS. 

Has anyone come across something similar ? Any thoughts would be greatly appreciated.

**Comment by Czaki** [comment]
I do not understand why but it fails with the building a wheel for `cffi` that provides a wheel for this python version.

**Reply by cheofusi** [reply]
Damn. Are you aware of any workarounds ?? Something like getting cibuildwheel to also build cffi ?

**Reply by Czaki** [reply]
Could you try to switch to PEP 517 build to declare build dependencies outside `setup.py`? 

https://github.com/zeromq/pyzmq

On the example page, I see 3 packages that declare that use CFFI https://cibuildwheel.readthedocs.io/en/stable/working-examples/

I'm pretty sure that some problems with CFFI were already mentioned in some solved issue/discussion. 

But the core problem is why it try to build cffi instead of download wheel.

**Reply by henryiii** [reply]
@Czaki is right. You are trying to use `setup_requires`, which is deprecated and doesn't support wheels, only the old (and no longer supported by pip) eggs: https://github.com/cheofusi/just_playback/blob/aaef83cc6dcc879d2cd2b08ac1949b47dcf924d7/setup.py#L33 - You need to remove that line and instead use a pyproject.toml:

```toml
[build-system]
requires = ["setuptools>=42", "cffi>=1.0.0"]
build-backend = "setuptools.build_meta"
```

This will solve your problem and enable local builds from SDists too.

**Reply by cheofusi** [reply]
Thank you both... I'll ditch setup.py as soon as I can

**Reply by henryiii** [reply]
You can't ditch setup.py, it's an extra file - you need setup.py and pyproject.toml. Your pyproject.toml file tells Python to read the setup.py and what dependencies it needs when it runs the setup.py. You can only remove this one line by adding that to pyproject.toml. (though you can technically remove quite a few of them with the new pyproject.toml support in setuptools 61+ if you want; I think you are only truly stuck with the `cffi_modules` one for now).

At least, not yet, but this got funded: https://iscinumpy.dev/post/scikit-build-proposal/, so I'm going to be working on better binary build options :)

**Reply by cheofusi** [reply]
Godspeed on scikit-build. Could one get away with pyproject.toml + setup.cfg + build ?

**Reply by henryiii** [reply]
Normally I'd say no, you can't make an Extension via setup.cfg or pyproject.toml. However, given that cffi has a setuptools plugin that just requires a configuration value, I'm not so sure; it _might_ be supported either in Setuptools' setup.cfg mode or pyproject.toml config mode.

**Reply by henryiii** [reply]
FYI, `pipx run hatch new --init` will auto-convert your setup.py into a hatchling based pyproject.toml, then you could just move back to setuptools and re-add this config value to `[tool.setuptools]`, that would be the quickest way to check setuptools 61+'s config mode.
